controller/volume_controller.go

```diff
@@ -904,17 +910,27 @@ func (vc *VolumeController) ReconcileVolumeState(v *longhorn.Volume, es map[stri
 			if !dataExists {
 				log.Warn("Cannot auto salvage volume: no data exists")
 			} else {
+				// This salvage is for revision counter enabled case
 				salvaged := false
+				// Since all replica failed, mark engine controller salvage requested
+				e.Spec.SalvageRequested = true
+				logrus.Infof("All replicas are failed, set engine salvageRequested to %v", e.Spec.SalvageRequested)
+
+				// Bring up the replicas for auto-salvage
 				for _, r := range failedUsableReplicas {
 					if util.TimestampWithinLimit(lastFailedAt, r.Spec.FailedAt, AutoSalvageTimeLimit) {
 						r.Spec.FailedAt = ""
-						log.WithField("replica", r.Name).Warn("Automatically salvaged volume replica")
-						msg := fmt.Sprintf("Replica %v of volume %v has been automatically salvaged", r.Name, v.Name)
-						vc.eventRecorder.Event(v, v1.EventTypeWarning, EventReasonAutoSalvaged, msg)
-						salvaged = true
+						log.WithField("replica", r.Name).Warn("Automatically salvaging volume replica")
+						// Revision counter enabled case, engine will
+						// check revision counter of replicas
+						if v.Spec.RevisionCounterDisabled {
+							msg := fmt.Sprintf("Replica %v of volume %v has been automatically salvaged", r.Name, v.Name)
+							vc.eventRecorder.Event(v, v1.EventTypeWarning, EventReasonAutoSalvaged, msg)
+							salvaged = true
+						}
 					}
 				}
-				if salvaged {
+				if salvaged || e.Spec.SalvageRequested {
```

Comment:
Due to line 916, the result of this check is always `true`. It is meaningless. 

I think you can remove line 926 `if v.Spec.RevisionCounterDisabled {` then revert (the modification of) this line.

Reply:
Sorry, missed `v.Spec.RevisionCounterDisabled` for `e.Spec.SalvageRequested`. Will update it. Thanks.

Reply:
Thinking about a corner case here, if there is no usable failed replica, `v.Status.PendingNodeID` will be still set as long as the revision counter is disabled. But actually, the re-attaching will get stuck there when there is no available replica. It's a little bit weird...